pkg/chains/storage/grafeas/grafeas.go

```diff
@@ -0,0 +1,384 @@
+		Attestation: &attestationpb.Details{
+			Attestation: &attestationpb.Attestation{
+				Signature: &attestationpb.Attestation_GenericSignedAttestation{
+					GenericSignedAttestation: &attestationpb.GenericSignedAttestation{
+						ContentType:       b.getContentType(opts),
+						SerializedPayload: payload,
+						Signatures: []*commonpb.Signature{
+							{
+								Signature:   []byte(signature),
+								PublicKeyId: b.cfg.Signers.KMS.KMSRef, // TODO: currently we only support storing kms keyID, will add other keys' ids later i.e. k8s secret, fulcio
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	envelope := &commonpb.Envelope{
+		Payload:     payload,
+		PayloadType: "in-toto attestations containing a slsa.dev/provenance predicate",
+		Signatures: []*commonpb.EnvelopeSignature{
+			{
+				Sig:   []byte(signature),
+				Keyid: b.cfg.Signers.KMS.KMSRef, // TODO: currently we only support storing kms keyID, will add other keys' ids later i.e. k8s secret, fulcio
+			},
+		},
+	}
+
+	var uri string
+	if opts.PayloadFormat == formats.PayloadTypeSimpleSigning {
+		uri = b.getOCIURI(opts)
+	} else {
+		uri = b.getTaskRunURI()
+	}
+
+	occurrence := &pb.Occurrence{
+		Resource: &pb.Resource{
+			Uri: uri,
+		},
+		NoteName: b.getNotePath(),
+		Details:  occurrenceDetails,
+		Envelope: envelope,
+	}
+
+	occurrenceRequest := &pb.CreateOccurrenceRequest{
+		Parent:     b.getProjectPath(),
+		Occurrence: occurrence,
+	}
+
+	return occurrenceRequest
+}
+
+func (b *Backend) getProjectPath() string {
+	projectID := b.cfg.Storage.Grafeas.ProjectID
+	return fmt.Sprintf(projectNameFormat, projectID)
+}
+
+func (b *Backend) getNotePath() string {
+	projectID := b.cfg.Storage.Grafeas.ProjectID
+	noteID := b.cfg.Storage.Grafeas.NoteID
+	return fmt.Sprintf(noteNameFormat, projectID, noteID)
+}
+
+// decide the attestation content type based on its format (simplesigning or in-toto)
+func (b *Backend) getContentType(opts config.StorageOpts) attestationpb.GenericSignedAttestation_ContentType {
+	// for simplesigning
+	if opts.PayloadFormat == formats.PayloadTypeSimpleSigning {
+		return attestationpb.GenericSignedAttestation_SIMPLE_SIGNING_JSON
+	}
+
+	// for in-toto
+	return attestationpb.GenericSignedAttestation_CONTENT_TYPE_UNSPECIFIED
+}
+
+// retrieve all occurrences created under a taskrun by filtering resource URI
+func (b *Backend) getOccurrences(ctx context.Context) ([]*pb.Occurrence, error) {
+	result := []*pb.Occurrence{}
+
+	// step 1: get all resource URIs created under the taskrun
+	uriFilters := []string{}
+	uriFilters = append(uriFilters, b.retrieveAllOCIURIs()...)
+	uriFilters = append(uriFilters, b.getTaskRunURI())
+
+	// step 2: find all occurrences by using ListOccurrences filters, and put them in result
+	for _, uriFilter := range uriFilters {
+		occs, err := b.findOccurrencesForCriteria(ctx, b.getProjectPath(), uriFilter)
+		if err != nil {
+			return result, err
+		}
+		result = append(result, occs...)
+	}
```

Comment:
This makes a ListRequest for every URI - could we construct a filter query that would fetch all of them in 1 request?

Reply:
Thanks for pointing out this! I refactored this by constructing a chained filter query first and then sending the request once!